File: flutter/lib/core/services/cleanup_service.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:supabase_flutter/supabase_flutter.dart';
// Edge functions invoked via Supabase client - no direct URL needed

/// Cleanup Service Provider
final cleanupServiceProvider = Provider<CleanupService>((ref) {
  return CleanupService();
});

/// Cleanup Service - Synced with React cleanup.service.ts
///
/// Triggers backend edge functions for data retention policy enforcement.
class CleanupService {
  final SupabaseClient _client = Supabase.instance.client;

  // Edge function calls use _client.functions.invoke() which handles URL/auth automatically

  /// Trigger data cleanup edge function
  Future<CleanupResult> triggerDataCleanup() async {
    try {
      final response = await _client.functions.invoke('data-cleanup');
      final data = response.data as Map<String, dynamic>?;
      return CleanupResult(
        success: data?['success'] ?? true,
        error: data?['error'] as String?,
      );
    } catch (e) {
      return CleanupResult(success: false, error: e.toString());
    }
  }

  /// Trigger group cleanup edge function
  Future<CleanupResult> triggerGroupCleanup() async {
    try {
      final response = await _client.functions.invoke('group-cleanup');
      final data = response.data as Map<String, dynamic>?;
      return CleanupResult(
        success: data?['success'] ?? true,
        error: data?['error'] as String?,
      );
    } catch (e) {
      return CleanupResult(success: false, error: e.toString());
    }
  }

  /// Trigger video cleanup edge function
  Future<CleanupResult> triggerVideoCleanup() async {
    try {
      final response = await _client.functions.invoke('video-cleanup');
      final data = response.data as Map<String, dynamic>?;
      return CleanupResult(
        success: data?['success'] ?? true,
        error: data?['error'] as String?,
      );
    } catch (e) {
      return CleanupResult(success: false, error: e.toString());
    }
  }

  /// Verify photo using edge function
  Future<PhotoVerificationResult> verifyPhoto({
    required String imageBase64,
    String? expectedGender,
    String? userId,
  }) async {
    try {
      final response = await _client.functions.invoke(
        'verify-photo',
        body: {
          'imageBase64': imageBase64,
          'expectedGender': expectedGender,
          'userId': userId,
          'verificationType': 'selfie',
        },
      );

      final data = response.data as Map<String, dynamic>?;
      return PhotoVerificationResult(
        verified: data?['verified'] ?? false,
        detectedGender: data?['detectedGender'] ?? data?['detected_gender'],
        confidence: (data?['confidence'] as num?)?.toDouble(),
        reason: data?['reason'] as String?,
        genderMatches: data?['genderMatches'] ?? data?['gender_matches'],
      );
    } catch (e) {
      // Accept photo on error to avoid blocking registration
      return PhotoVerificationResult(verified: true, reason: 'Photo accepted');
    }
  }

  /// Run all cleanup tasks (admin manual trigger)
  Future<AllCleanupResults> runAllCleanups() async {
    final results = await Future.wait([
      triggerDataCleanup(),
      triggerGroupCleanup(),
      triggerVideoCleanup(),
    ]);

    return AllCleanupResults(
      dataCleanup: results[0],
      groupCleanup: results[1],
      videoCleanup: results[2],
    );
  }
}

/// Result model for cleanup operations
class CleanupResult {
  final bool success;
  final String? error;

  CleanupResult({required this.success, this.error});
}

/// Result model for photo verification
class PhotoVerificationResult {
  final bool verified;
  final String? detectedGender;
  final double? confidence;
  final String? reason;
  final bool? genderMatches;

  PhotoVerificationResult({
    required this.verified,
    this.detectedGender,
    this.confidence,
    this.reason,
    this.genderMatches,
  });
}

/// Combined results for all cleanup operations
class AllCleanupResults {
  final CleanupResult dataCleanup;
  final CleanupResult groupCleanup;
  final CleanupResult videoCleanup;

  AllCleanupResults({
    required this.dataCleanup,
    required this.groupCleanup,
    required this.videoCleanup,
  });
}

File: src/services/cleanup.service.ts
import { classifyError, logError } from "@/lib/errors";
import { supabase } from "@/integrations/supabase/client";

/**
 * Triggers the data-cleanup edge function to clean up old data
 */
export const triggerDataCleanup = async (): Promise<{ success: boolean; error?: string }> => {
  try {
    const { data, error } = await supabase.functions.invoke('data-cleanup', {
      body: {},
    });

    if (error) throw error;
    return { success: data?.success ?? true, error: data?.error };
  } catch (error) {
    console.error('Data cleanup error:', error);
    return { success: false, error: classifyError(error, 'run cleanup').message };
  }
};

/**
 * Triggers the group-cleanup edge function to clean up old group messages
 */
export const triggerGroupCleanup = async (): Promise<{ success: boolean; error?: string }> => {
  try {
    const { data, error } = await supabase.functions.invoke('group-cleanup', {
      body: {},
    });

    if (error) throw error;
    return { success: data?.success ?? true, error: data?.error };
  } catch (error) {
    console.error('Group cleanup error:', error);
    return { success: false, error: error instanceof Error ? error.message : 'Unknown error' };
  }
};

/**
 * Triggers the video-cleanup edge function to clean up old video sessions
 */
export const triggerVideoCleanup = async (): Promise<{ success: boolean; error?: string }> => {
  try {
    const { data, error } = await supabase.functions.invoke('video-cleanup', {
      body: {},
    });

    if (error) throw error;
    return { success: data?.success ?? true, error: data?.error };
  } catch (error) {
    console.error('Video cleanup error:', error);
    return { success: false, error: error instanceof Error ? error.message : 'Unknown error' };
  }
};

/**
 * Verifies a photo using the verify-photo edge function
 */
export const verifyPhoto = async (
  imageBase64: string,
  expectedGender?: 'male' | 'female',
  userId?: string
): Promise<{
  verified: boolean;
  detectedGender?: string;
  confidence?: number;
  reason?: string;
  genderMatches?: boolean;
}> => {
  try {
    const { data, error } = await supabase.functions.invoke('verify-photo', {
      body: {
        imageBase64,
        expectedGender,
        userId,
        verificationType: 'selfie',
      },
    });

    if (error) throw error;

    return {
      verified: data?.verified ?? false,
      detectedGender: data?.detectedGender ?? data?.detected_gender,
      confidence: data?.confidence,
      reason: data?.reason,
      genderMatches: data?.genderMatches ?? data?.gender_matches,
    };
  } catch (error) {
    console.error('Photo verification error:', error);
    return { verified: false, reason: 'Photo verification failed. Please try again.' };
  }
};

/**
 * Runs all cleanup tasks - useful for admin manual trigger
 */
export const runAllCleanups = async (): Promise<{
  dataCleanup: { success: boolean; error?: string };
  groupCleanup: { success: boolean; error?: string };
  videoCleanup: { success: boolean; error?: string };
}> => {
  const [dataCleanup, groupCleanup, videoCleanup] = await Promise.all([
    triggerDataCleanup(),
    triggerGroupCleanup(),
    triggerVideoCleanup(),
  ]);

  return { dataCleanup, groupCleanup, videoCleanup };
};
